ls1intum/Artemis
# `Programming exercises`: Distinguish test case types into structural and behavioral in Java exercises

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java. In order to generate code hints, we have to distinguish between structural and behavioral test cases because we will generate hints for both types in a different way.
(Side note: We will string-generate hints for structural tests while we use coverage results for behavioral tests.)

### Description
This PR introduce a new enumeration `ProgrammingExerciseTestCaseType` that is an attribute of `ProgrammingExerciseTestCase`. There are three different values of this enumeration:
- STRUCTURAL: Java test cases, that have been generated by the structure oracle and test the code for structural aspects as classes, constructors, methods, and attributes.
- BEHAVIORAL: All remaining Java test cases. These are all test cases, that have been written by an editor (manually).
- DEFAULT: Test cases in all remaining programming languages. Since our Bachelor Thesis focuses on code hint generation for Java code only, the type for test cases in all other programming languages are set to this value. This however can be changed when support for other programming languages would be introduced.

The test cases are mapped by their name. Structural test cases in Java in Artemis match the following pattern:
- "testClass[\<className\>]": Structural tests for a class
- "testConstructors[\<className\>]": Structural tests for constructors
- "testMethods[\<methodName\>]": Structural tests for methods
- "testAttributes[\<attributeName\>]": Structural tests for attributes
All remaining Java test cases are of type BEHAVIORAL.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Programming Exercise in Java with structural and behavioral tests (are contained in the template exercises)
- 1 Programming Exercise in any language other than Java.

1. Log in to Artemis
2. Navigate to Course Overview -> Exercises
3. Click on a programming exercise
4. Open the grading view to display all exercises with the corresponding test case type. Check that the types are only displayed for Java exercises. Also check that the types are correctly (as described above) for each of the following:
    - newly created programming exercises
    - imported programming exercises
    - existing and unchanged programming exercises
    - programming exercises where structural and behavioral test cases are changed

**Important note**: In the following scenario the types for test cases are set to DEFAULT, when they should be either STRUCTURAL or BEHAVIORAL:

0. Some programming exercises exists.
1. This PR has been deployed the first time to a TS and the (Java) migration setting the type for existing test cases is performed. It will be saved in the database that this migration has been performed.
2. Any other PR is deployed to this TS.
3. A JAVA programming exercise is created. The type for all test cases will be set to the standard value DEFAULT.
4. This PR is deployed again. Because the migration has already been performed, it will not be run again. The test case types for the programming exercises from 3. will therefore not be updated and remain DEFAULT.
--> This behavior would obviously not occur when deployed to the live server, because when the PR is deployed once, the migration will be executed once and the types will be set for every exercise that is created from then on.


### Review Progress
#### Code Review
#### Manual Tests

### Screenshots
<img width="827" alt="Bildschirmfoto 2022-02-21 um 10 51 31" src="https://user-images.githubusercontent.com/73833780/154930679-afeb21b7-ceb3-41d4-a965-1832b8f6219e.png">
